results/cmd/watcher/main_test.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"context"
+	"net"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	ttesting "github.com/tektoncd/pipeline/pkg/reconciler/testing"
+	test "github.com/tektoncd/pipeline/test"
+
+	_ "github.com/mattn/go-sqlite3"
+	"github.com/tektoncd/experimental/results/pkg/server"
+	pb "github.com/tektoncd/experimental/results/proto/proto"
+	"google.golang.org/grpc"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/configmap"
+)
+
+const (
+	port = ":50051"
+)
+
+func getRunName(tr *v1beta1.TaskRun) string {
+	return strings.Join([]string{tr.Namespace, tr.Name}, "/")
+}
+
+// getFakeClients create a fake client to send test data to reconciler
+func getFakeClients(t *testing.T, tr []*v1beta1.TaskRun, conn *grpc.ClientConn) (test.Assets, context.CancelFunc) {
+	t.Helper()
+	ctx, _ := ttesting.SetupFakeContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	d := test.Data{
+		TaskRuns: tr,
+	}
+	clients, _ := test.SeedTestData(t, ctx, d)
+	cmw := configmap.NewInformedWatcher(clients.Kube, "")
+	return test.Assets{
+		Controller: newController(ctx, cmw, conn),
+		Clients:    clients,
+	}, cancel
+}
+
+// reconcileTaskRun sends TaskRun data to reconciler and then retrieves completed TaskRun
+func reconcileTaskRun(taskRun v1beta1.TaskRun, d []*v1beta1.TaskRun, t *testing.T, conn *grpc.ClientConn) (*v1beta1.TaskRun, error) {
+	testAssets, cancel := getFakeClients(t, d, conn)
+	defer cancel()
+	c := testAssets.Controller
+	clients := testAssets.Clients
+	ctx := context.Background()
+	if err := c.Reconciler.Reconcile(ctx, getRunName(&taskRun)); err != nil {
+		return nil, err
+	}
+	tr, err := clients.Pipeline.TektonV1beta1().TaskRuns(taskRun.Namespace).Get(taskRun.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return tr, nil
+}
+
+// TestReconcile tests if TaskRun in the client can be properly updated when sent to reconciler
+func TestReconcile(t *testing.T) {
+	srv, err := server.SetupTestDB(t)
+	if err != nil {
+		t.Fatalf("Failed to create fake server: %v", err)
+	}
+	s := grpc.NewServer()
+	pb.RegisterResultsServer(s, srv) // local test server
+	lis, err := net.Listen("tcp", port)
+	if err != nil {
+		//
+	}
+	defer lis.Close()
+	go s.Serve(lis)
+	defer s.Stop()
+	conn, err := grpc.Dial(*apiAddr, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		t.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+	taskRun := v1beta1.TaskRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "Tekton",
+			Namespace:   "default",
+			Annotations: map[string]string{"demo": "demo"},
+		},
+	}
+	// Create a new TaskRun in the fake client, completed TaskRun should be updated with a results id field in annotations.
+	tr, err := reconcileTaskRun(taskRun, []*v1beta1.TaskRun{&taskRun}, t, conn)
```

Comment:
I think @wlynch already mentioned this. But we are calling reconcileTaskRun twice in this function. And inside that function, we are initializing the fake clients and the controller each time. Which means that there are two sets of controllers and clients. What we should do instead is call `getFakeClients` once, get the controller from the returned testAssets and then call c.Reconciler.Reconcile twice. In other words, we do the setup only once but call Reconcile twice to verify we are setting the results id annotation.

Reply:
I understand that functionally this won't be too different from what we have now because in the second call to `reconcileTaskRun`, we are seeding the fake client with a Taskrun which already has a results_id set. What this will help with is clarity and readability i.e. when I read the test, it becomes easier for me to understand what is going on. Its like there are two test cases here in the same test:

1. We check that when we reconcile a TaskRun that has no results_id set, we set the results_id. i.e we seed fake clients with the initial taskrun, call reconcile, and then we can get the taskrun to check that it has a results_id annotation.

2. We check that when we reconcile a TaskRun that already has a results_id set, we do not set a new results_id but use the already set results_id.

I think it might be clearer to have two tests (or two subtests in the same test) to make it clearer what is going on.

Reply:
Thanks for the suggestion, for the second one, I tried to use a pre-set results_id, the issue is for UpdateTaskRun(), if it cannot find a taskrun with specified results_id, it will return an error, so the test with a hardcoding TaskRun data will immediately fail. This is the reason why I use a TaskRun created before.

Reply:
I agree with what you said, I think I can create just a new client instead of both client and controller for the two consecutive tests.

Reply:
done